DevEdu.API/DevEdu.Business.Tests/UserServiceTests.cs

```diff
@@ -74,19 +96,247 @@ public void UpdateUser_UserDto_ReturnUpdateUserDto()
+            _repoMock.Setup(x => x.AddUserRole(UserData.expectedUserId, It.IsAny<int>()));
+
+            //When
+            var actualId = _userService.AddUser(user);
+
+            //Then
+            Assert.AreEqual(UserData.expectedUserId, actualId);
+            _repoMock.Verify(x => x.AddUser(user), Times.Once);
+            _repoMock.Verify(x => x.AddUserRole(actualId, It.IsAny<int>()), Times.AtLeastOnce);
+        }
+
+
+        [TestCase(0)]
+        [TestCase(-100)]
+        public void SelectUserById_WhenIdLessThanAllowedValue_Exception(int id)
+        {
+            //Given
+            var expectedException = string.Format(ServiceMessages.MinimumAllowedValueMessage, nameof(id), UserData.idMinimum);
+
+            //When
+            var ex = Assert.Throws<Exception>(
+                () => _userService.SelectUserById(id));
+
+            //Then
+            Assert.That(ex.Message, Is.EqualTo(expectedException));
+            _repoMock.Verify(x => x.SelectUserById(id), Times.Never);
+        }
+
+        [Test]
+        public void SelectUserByEmail_WhenDoNotHaveMatchesInDataBase_EntityNotFoundException()
+        {
+            //Given
+            var user = UserData.GetUserDto();
+            var expectedException = $"{nameof(user)} with email = {user.Email} was not found";
+
+            //When
+            var ex = Assert.Throws<EntityNotFoundException>(
+                () => _userService.SelectUserByEmail(user.Email));
+
+            //Then
+            Assert.That(ex.Message, Is.EqualTo(expectedException));
+            _repoMock.Verify(x => x.SelectUserByEmail(user.Email), Times.Once);
+        }
+
+        [Test]
+        public void SelectUsers_WhenListIsEmpty_Exception()
+        {
+            //Given
+            var list = _repoMock.Setup(x => x.SelectUsers()).Returns(new List<UserDto>() { });
+            var expectedException = $"{nameof(list)} is empty";
+
+            //When
+            var ex = Assert.Throws<Exception>(
+                () => _userService.SelectUsers());
+
+            //Then
+            Assert.That(ex.Message, Is.EqualTo(expectedException));
+            _repoMock.Verify(x => x.SelectUsers(), Times.Once);
+        }
+
+        [Test]
+        public void UpdateUser_WhenDoNotHaveMatchesInDataBase_EntityNotFoundException()
+        {
+            //Given
+            var user = UserData.GetUserDto();
+            var expectedException = string.Format(ServiceMessages.EntityNotFoundMessage, nameof(user), user.Id);
+
+            //When
+            var ex = Assert.Throws<EntityNotFoundException>(
+                () => _userService.UpdateUser(user));
+
+            //Then
+            Assert.That(ex.Message, Is.EqualTo(expectedException));
+            _repoMock.Verify(x => x.SelectUserById(user.Id), Times.Once);
+            _repoMock.Verify(x => x.UpdateUser(user), Times.Never);
+        }
+
+        [TestCase(1)]
+        [TestCase(100)]
+        [TestCase(100500)]
```

Comment:
usually, we don't use test cases in such tests
only one test per case is enough